### ISSUE TITLE
Sub-Model used as layer does not support masking

### ISSUE THREAD

**Issue by taivanbat** [fchollet/keras#6541]
Using a sub-model with masking like the following:

```
# make encoder
encoder_input = Input(shape=(MAX_SENTENCE_SIZE, VOCAB_SIZE),
                          name='encoder_input')
# masking = Masking(mask_value=0.0)(encoder_input)

encoded = LSTM(ENCODING_SIZE, return_sequences=False, name='encoded')(encoder_input)
encoder = Model(encoder_input, encoded)

# make decoder model
decoder_input = Input(shape=(ENCODING_SIZE,))
decoded = RepeatVector(MAX_SENTENCE_SIZE)(decoder_input)
decoded = LSTM(VOCAB_SIZE, return_sequences=True)(decoded)
decoded = TimeDistributed(Dense(VOCAB_SIZE, activation='softmax',
                                    name='decoded'))(decoded)

decoder = Model(decoder_input, decoded)

# make sequence autoencoder
encoder_decoder_input = Input(shape=(MAX_SENTENCE_SIZE, VOCAB_SIZE),
                              name='encoder_decoder_input')
encoder_output = encoder(encoder_decoder_input)
decoder_output = decoder(encoder_output)

sequence_autoencoder = Model(inputs=encoder_decoder_input, outputs=decoder_output)
sequence_autoencoder.compile(loss='categorical_crossentropy', optimizer='adam')

encoder.compile(loss='categorical_crossentropy', optimizer='adam')
```

But this gives an error: 

```
[[Node: encoder_input = Placeholder[dtype=DT_FLOAT, shape=[], _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
W tensorflow/core/framework/op_kernel.cc:993] Invalid argument: You must feed a value for placeholder tensor 'encoder_input' with dtype float
	 [[Node: encoder_input = Placeholder[dtype=DT_FLOAT, shape=[], _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
```

Removing the masking layer fixes the issue. How do I mask the inputs here? I have tried using a masking layer at the sequence_autoencoder input but then the Input layer for the encoder does not accept a value of type bool.

**Comment by taivanbat** [comment]
#3524 has the same issue but it has not been solved, it seems.

**Comment by brooksjd** [comment]
I've noticed the same issue, and get the same tensorflow error. It seems like if you define the entire model with the functional API it will work.
